### PR TITLE
Implementation of telegram using netx

### PR DESCRIPTION
**Pull request by bassosimone**
This PR started as an attempt to write a minimal implementation of Telegram based on netx and later evolved to encompass processing of netx events and code testing.

### CODE REVIEW

File: experiment/experiment.go [hunk 1]
```diff
@@ -142,7 +142,10 @@ func (e *Experiment) Measure(
 		return
 	}
 	measurement = e.newMeasurement(input)
+	start := time.Now()
 	err = e.DoMeasure(ctx, e.Session, &measurement, e.Callbacks)
+	stop := time.Now()
+	measurement.MeasurementRuntime = stop.Sub(start).Seconds()
```
Comment (hellais) [inline experiment/experiment.go]: This is cool because:
> this code always computes a positive elapsed time, even if the wall clock is changed during the operation being timed

File: experiment/telegram/telegram_otherwise.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	"github.com/ooni/probe-engine/experiment"
+	"github.com/ooni/probe-engine/experiment/handler"
+	"github.com/ooni/probe-engine/log"
+	"github.com/ooni/probe-engine/model"
+	"github.com/ooni/probe-engine/session"
+)
+
+const (
+	testName    = "telegram"
+	testVersion = "0.0.2"
```
Comment (hellais) [inline experiment/telegram/telegram_otherwise.go]: I suggest bumping this up to a number different than version being used in measurement-kit so we can more easily distinguish them.

File: experiment/telegram/telegram_otherwise.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	m.tk.TelegramTCPBlocking = true
+	m.tk.TelegramHTTPBlocking = true
+	var addresses = []string{
+		"149.154.175.50", "149.154.167.51", "149.154.175.100",
+		"149.154.167.91", "149.154.171.5",
+	}
+	for _, addr := range addresses {
+		for _, port := range []string{"80", "443"} {
+			const alwaysHTTP = "http" // note: it's intended to use HTTP with 443
+			resp, err := m.request(ctx, "POST", alwaysHTTP, addr, port)
```
Comment (hellais) [inline experiment/telegram/telegram_otherwise.go]: Is doing this in fact equivalent to just doing a TCP connection to the IP and port?

Comment (hellais) [reply experiment/telegram/telegram_otherwise.go]: I suspect that if the connection is aborted or reset after you send the HTTP POST payload you will still fall into the `if err != nil` below and result in marking the result as `tcp_blocking`, while in fact it's a different scenario.

File: experiment/telegram/telegram_otherwise.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+// Config contains the experiment config.
+type Config struct{}
+
+// TestKeys contains telegram test keys.
+type TestKeys struct {
+	TelegramHTTPBlocking bool   `json:"telegram_http_blocking"`
+	TelegramTCPBlocking  bool   `json:"telegram_tcp_blocking"`
+	TelegramWebFailure   string `json:"telegram_web_failure"`
+	TelegramWebStatus    string `json:"telegram_web_status"`
```
Comment (hellais) [inline experiment/telegram/telegram_otherwise.go]: Where is the `tcp_connect` list populated with the result items?

**Comment by hellais** [comment]
Hum, I see that many fields are missing from the output of this test.

```
% python -m json.tool report-netx.jsonl
{
    "data_format_version": "0.2.1",
    "measurement_start_time": "2019-10-09 13:11:42",
    "test_runtime": 2.238912705,
    "probe_asn": "AS30722",
    "probe_cc": "IT",
    "probe_ip": "127.0.0.1",
    "report_id": "20191009T131150Z_AS30722_XlgNi0CXrnHPomrNbkanhK556nuCCl94Jlw9p0W1RRhc2aV7xM",
    "resolver_ip": "172.253.197.4",
    "software_name": "miniooni",
    "software_version": "0.1.0-dev",
    "test_keys": {
        "telegram_http_blocking": false,
        "telegram_tcp_blocking": false,
        "telegram_web_failure": "",
        "telegram_web_status": "ok"
    },
    "test_name": "telegram",
    "test_start_time": "2019-10-09 13:11:42",
    "test_version": "0.0.2"
}
```

Here it is with measurement-kit

```
 % python -m json.tool report-measurement-kit.jsonl
{
    "annotations": {
        "engine_name": "libmeasurement_kit",
        "engine_version": "0.10.6",
        "engine_version_full": "v0.10.6",
        "platform": "macos"
    },
    "data_format_version": "0.2.0",
    "id": "cce9601c-bde7-4ac5-acdc-6125881caece",
    "measurement_start_time": "2019-10-09 13:07:55",
    "test_runtime": 1.279407221,
    "probe_asn": "AS30722",
    "probe_cc": "IT",
    "probe_ip": "127.0.0.1",
    "report_id": "20191009T130803Z_AS30722_6Cf2mgwXS0wp8Ha4rZ1fPnHczSobjUCMDykY45FsUqMq5RwfXA",
    "resolver_ip": "172.253.199.4",
    "software_name": "miniooni",
    "software_version": "0.1.0-dev",
    "test_keys": {
        "agent": "agent",
        "client_resolver": "127.0.0.1",
        "requests": [
            {
                "failure": null,
                "request": {
                    "body": "",
                    "headers": {
                        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                        "Accept-Language": "en-US;q=0.8,en;q=0.5",
                        "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                    },
                    "method": "GET",
                    "tor": {
                        "exit_ip": null,
                        "exit_name": null,
                        "is_tor": false
                    },
                    "url": "http://web.telegram.org/"
                },
                "response": {
                    "body": "<!doctype html><html lang=en manifest=webogram.appcache ng-csp xmlns:ng=http://angularjs.org id=ng-app style=\"display: none;\"><head><meta charset=utf-8><meta name=viewport content=\"width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no\"><title>Telegram Web</title><link rel=stylesheet href=css/app.css><link rel=manifest href=manifest.webapp.json><link rel=icon href=favicon.ico type=image/x-icon><link rel=apple-touch-icon href=img/iphone_home120.png><link rel=apple-touch-icon sizes=120x120 href=img/iphone_home120.png><link rel=apple-touch-startup-image media=\"(device-width: 320px)\" href=img/iphone_startup.png><meta name=apple-mobile-web-app-title content=\"Telegram Web\"><meta name=mobile-web-app-capable content=yes><meta name=apple-mobile-web-app-capable content=yes><meta name=apple-mobile-web-app-status-bar-style content=black-translucent><meta name=theme-color content=#497495><meta name=google content=notranslate><meta property=og:title content=\"Telegram Web\"><meta property=og:url content=\"https://web.telegram.org/\"><meta property=og:image content=https://web.telegram.org/img/logo_share.png><meta property=og:site_name content=\"Telegram Web\"><meta property=description content=\"Welcome to the Web application of Telegram messenger. See https://github.com/zhukov/webogram for more info.\"><meta property=og:description content=\"Welcome to the Web application of Telegram messenger. See https://github.com/zhukov/webogram for more info.\"></head><body><div class=page_wrap ng-view></div><div id=notify_sound></div><script src=js/app.js></script></body></html>",
                    "code": 200,
                    "headers": {
                        "Accept-Ranges": "bytes",
                        "Cache-Control": "max-age=3600",
                        "Connection": "keep-alive",
                        "Content-Length": "1587",
                        "Content-Type": "text/html",
                        "Date": "Wed, 09 Oct 2019 13:08:04 GMT",
                        "ETag": "\"5d83f6a8-633\"",
                        "Expires": "Wed, 09 Oct 2019 14:08:04 GMT",
                        "Last-Modified": "Thu, 19 Sep 2019 21:44:08 GMT",
                        "Server": "nginx/1.16.1"
                    },
                    "response_line": "HTTP/1.1 200 OK"
                }
            }, [..snip..]
            }
        ],
        "socksproxy": null,
        "tcp_connect": [
            {
                "ip": "149.154.167.51",
                "port": 80,
                "status": {
                    "failure": null,
                    "success": true
                }
            }, [.. snip ..]
            }
        ],
        "telegram_http_blocking": false,
        "telegram_tcp_blocking": false,
        "telegram_web_failure": null,
        "telegram_web_status": "ok"
    },
    "test_name": "telegram",
    "test_start_time": "2019-10-09 13:07:55",
    "test_version": "0.0.2"
}
```

File: experiment/telegram/telegram_otherwise.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+)
+
+// Config contains the experiment config.
+type Config struct{}
+
+// TestKeys contains telegram test keys.
+type TestKeys struct {
+	TelegramHTTPBlocking bool   `json:"telegram_http_blocking"`
+	TelegramTCPBlocking  bool   `json:"telegram_tcp_blocking"`
+	TelegramWebFailure   string `json:"telegram_web_failure"`
```
Comment (hellais) [inline experiment/telegram/telegram_otherwise.go]: It should be noted that we are changing the semantics of this key. In the current implementation it is `null` when there is no failure, while now it's an empty string.

If you decide to change it be sure to also update the test specification.

File: experiment/experiment.go (hunk 1 above)
Comment (bassosimone) [reply experiment/experiment.go]: Yes, a number of techniques are being implemented in `netx` and here. Also, this is a bug fix that I noticed when understanding the output of this experiment :).

File: experiment/telegram/telegram_otherwise.go (hunk 2 above)
Comment (bassosimone) [reply experiment/telegram/telegram_otherwise.go]: Yes, good suggestion.

File: experiment/telegram/telegram_otherwise.go (hunk 4 above)
Comment (bassosimone) [reply experiment/telegram/telegram_otherwise.go]: It is not. As mentioned in the PR description, this is a minimal implementation of Telegram that does not include any data format like `tcp_connect`. Those should be computed using the data stream from `netx` using the same code for all tests. There is no point in adding them in this stage of the PR, except making the diff long and difficult to read.

Comment (bassosimone) [reply experiment/telegram/telegram_otherwise.go]: Correction: there was no point for me to adding them at the beginning of the review of the PR except making the diff more long and difficult to read for you. That said, I am open to add this code in here if you wish. My preference, however, would be to do that at a later stage when we have a bunch of tests, so I can better understand what is required.

File: experiment/telegram/telegram_otherwise.go (hunk 5 above)
Comment (bassosimone) [reply experiment/telegram/telegram_otherwise.go]: Good catch, this is a mistake. I'd rather not change the spec.

File: experiment/telegram/telegram_otherwise.go (hunk 4 above)
Comment (hellais) [reply experiment/telegram/telegram_otherwise.go]: Hum, I think it would be useful to have it done as part of this PR so that we can better understand how everything fits together and validate that the result of this test matches up with the result of the measurement-kit based one (i.e. I can setup some rules on my machine to simulate some forms of blocking and check that both tests produce similar results).

Beyond just looking at the diff I think it's most useful to actually test the code and compare the two code-bases against each other.

File: experiment/telegram/telegram_otherwise.go (hunk 2 above)
Comment (bassosimone) [reply experiment/telegram/telegram_otherwise.go]: done!

File: experiment/telegram/telegram_otherwise.go (hunk 5 above)
Comment (bassosimone) [reply experiment/telegram/telegram_otherwise.go]: done!

File: experiment/telegram/telegram_otherwise.go (hunk 3 above)
Comment (bassosimone) [reply experiment/telegram/telegram_otherwise.go]: So, does `tcp_blocking` in this context mean only failure to connect and everything else is HTTP blocking?

File: experiment/telegram/telegram_otherwise.go (hunk 4 above)
Comment (bassosimone) [reply experiment/telegram/telegram_otherwise.go]: As discussed in video chat, this functionality will be added later and we will proceed piecemeal for now.

**Comment by bassosimone** [comment]
@hellais I will clarify the scope of the review, as discussed in videochat. I would like to proceed piecemeal and keep PRs small and quick to review. This just aims to add the skeleton. A subsequent PR will start using netx events to populate more keys. Work will be done when we will have a fully tested implementation with tests matching edge cases we care about.

File: experiment/telegram/telegram_otherwise.go (hunk 4 above)
Comment (bassosimone) [reply experiment/telegram/telegram_otherwise.go]: ok, will implement the full test then

**Comment by bassosimone** [comment]
@hellais I have implemented the whole test, as requested. I left and clearly marked some parts of the code as _TODO soon_ because they're out of scope for this PR. Please take another look!

```JSON

{
  "data_format_version": "0.2.1",
  "measurement_start_time": "2019-10-10 19:04:12",
  "test_runtime": 2.715765107,
  "probe_asn": "AS30722",
  "probe_cc": "IT",
  "probe_ip": "127.0.0.1",
  "report_id": "20191010T190412Z_AS30722_cUCJkzcB0rFNCkMdjDeEoKOwiZrlZZe13Za70k3XYRJTSTdoG7",
  "resolver_ip": "172.253.199.4",
  "software_name": "miniooni",
  "software_version": "0.1.0-dev",
  "test_keys": {
    "requests": [
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Host": "web.telegram.org",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "GET",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "https://web.telegram.org"
        },
        "response": {
          "body": "<!doctype html><html lang=en manifest=webogram.appcache ng-csp xmlns:ng=http://angularjs.org id=ng-app style=\"display: none;\"><head><meta charset=utf-8><meta name=viewport content=\"width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no\"><title>Telegram Web</title><link rel=stylesheet href=css/app.css><link rel=manifest href=manifest.webapp.json><link rel=icon href=favicon.ico type=image/x-icon><link rel=apple-touch-icon href=img/iphone_home120.png><link rel=apple-touch-icon sizes=120x120 href=img/iphone_home120.png><link rel=apple-touch-startup-image media=\"(device-width: 320px)\" href=img/iphone_startup.png><meta name=apple-mobile-web-app-title content=\"Telegram Web\"><meta name=mobile-web-app-capable content=yes><meta name=apple-mobile-web-app-capable content=yes><meta name=apple-mobile-web-app-status-bar-style content=black-translucent><meta name=theme-color content=#497495><meta name=google content=notranslate><meta property=og:title content=\"Telegram Web\"><meta property=og:url content=\"https://web.telegram.org/\"><meta property=og:image content=https://web.telegram.org/img/logo_share.png><meta property=og:site_name content=\"Telegram Web\"><meta property=description content=\"Welcome to the Web application of Telegram messenger. See https://github.com/zhukov/webogram for more info.\"><meta property=og:description content=\"Welcome to the Web application of Telegram messenger. See https://github.com/zhukov/webogram for more info.\"></head><body><div class=page_wrap ng-view></div><div id=notify_sound></div><script src=js/app.js></script></body></html>",
          "code": 200,
          "headers": {
            "Cache-Control": "max-age=3600",
            "Connection": "keep-alive",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:15 GMT",
            "Etag": "W/\"5d83f6a8-633\"",
            "Expires": "Thu, 10 Oct 2019 20:04:15 GMT",
            "Last-Modified": "Thu, 19 Sep 2019 21:44:08 GMT",
            "Server": "nginx/1.16.1",
            "Strict-Transport-Security": "max-age=35768000"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Host": "web.telegram.org",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "GET",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://web.telegram.org"
        },
        "response": {
          "body": "<!doctype html><html lang=en manifest=webogram.appcache ng-csp xmlns:ng=http://angularjs.org id=ng-app style=\"display: none;\"><head><meta charset=utf-8><meta name=viewport content=\"width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no\"><title>Telegram Web</title><link rel=stylesheet href=css/app.css><link rel=manifest href=manifest.webapp.json><link rel=icon href=favicon.ico type=image/x-icon><link rel=apple-touch-icon href=img/iphone_home120.png><link rel=apple-touch-icon sizes=120x120 href=img/iphone_home120.png><link rel=apple-touch-startup-image media=\"(device-width: 320px)\" href=img/iphone_startup.png><meta name=apple-mobile-web-app-title content=\"Telegram Web\"><meta name=mobile-web-app-capable content=yes><meta name=apple-mobile-web-app-capable content=yes><meta name=apple-mobile-web-app-status-bar-style content=black-translucent><meta name=theme-color content=#497495><meta name=google content=notranslate><meta property=og:title content=\"Telegram Web\"><meta property=og:url content=\"https://web.telegram.org/\"><meta property=og:image content=https://web.telegram.org/img/logo_share.png><meta property=og:site_name content=\"Telegram Web\"><meta property=description content=\"Welcome to the Web application of Telegram messenger. See https://github.com/zhukov/webogram for more info.\"><meta property=og:description content=\"Welcome to the Web application of Telegram messenger. See https://github.com/zhukov/webogram for more info.\"></head><body><div class=page_wrap ng-view></div><div id=notify_sound></div><script src=js/app.js></script></body></html>",
          "code": 200,
          "headers": {
            "Cache-Control": "max-age=3600",
            "Connection": "keep-alive",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:15 GMT",
            "Etag": "W/\"5d83f6a8-633\"",
            "Expires": "Thu, 10 Oct 2019 20:04:15 GMT",
            "Last-Modified": "Thu, 19 Sep 2019 21:44:08 GMT",
            "Server": "nginx/1.16.1"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.171.5:443",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.171.5:443"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:15 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.171.5:80",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.171.5:80"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:14 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.167.91:443",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.167.91:443"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:14 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.167.91:80",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.167.91:80"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:14 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.175.100:443",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.175.100:443"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:14 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.175.100:80",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.175.100:80"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:13 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.167.51:443",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.167.51:443"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:13 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.167.51:80",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.167.51:80"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:13 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.175.50:443",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.175.50:443"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:13 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      },
      {
        "failure": null,
        "request": {
          "body": "",
          "headers": {
            "Accept-Encoding": "gzip",
            "Content-Length": "0",
            "Host": "149.154.175.50:80",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
          },
          "method": "POST",
          "tor": {
            "exit_ip": null,
            "exit_name": null,
            "is_tor": false
          },
          "url": "http://149.154.175.50:80"
        },
        "response": {
          "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
          "code": 501,
          "headers": {
            "Content-Length": "181",
            "Content-Type": "text/html",
            "Date": "Thu, 10 Oct 2019 19:04:13 GMT",
            "Server": "nginx/0.3.33"
          }
        }
      }
    ],
    "tcp_connect": [
      {
        "ip": "149.154.175.50",
        "port": 80,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.175.50",
        "port": 443,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.167.51",
        "port": 80,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.167.51",
        "port": 443,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.175.100",
        "port": 80,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.175.100",
        "port": 443,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.167.91",
        "port": 80,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.167.91",
        "port": 443,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.171.5",
        "port": 80,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.171.5",
        "port": 443,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.167.99",
        "port": 80,
        "status": {
          "failure": null,
          "success": true
        }
      },
      {
        "ip": "149.154.167.99",
        "port": 443,
        "status": {
          "failure": null,
          "success": true
        }
      }
    ],
    "telegram_http_blocking": false,
    "telegram_tcp_blocking": false,
    "telegram_web_failure": null,
    "telegram_web_status": "ok"
  },
  "test_name": "telegram",
  "test_start_time": "2019-10-10 19:04:12",
  "test_version": "0.0.3"
}
```

**Comment by hellais** [comment]
I am comparing the output of this test with the current telegram test in MK and I see that for all the HTTP requests the response are different:

`mk`
```
            {
                "failure": null,
                "request": {
                    "body": "",
                    "headers": {
                        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                        "Accept-Language": "en-US;q=0.8,en;q=0.5",
                        "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                    },
                    "method": "POST",
                    "tor": {
                        "exit_ip": null,
                        "exit_name": null,
                        "is_tor": false
                    },
                    "url": "http://149.154.175.100:443/"
                },
                "response": {
                    "body": "<html>\r\n<head><title>411 Length Required</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>411 Length Required</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
                    "code": 411,
                    "headers": {
                        "Connection": "keep-alive",
                        "Content-Length": "181",
                        "Content-Type": "text/html",
                        "Date": "Mon, 14 Oct 2019 15:28:53 GMT",
                        "Server": "nginx/0.3.33"
                    },
                    "response_line": "HTTP/1.1 411 Length Required"
                }
            },
```


`netx`
```
            {
                "failure": null,
                "request": {
                    "body": "",
                    "headers": {
                        "Accept-Encoding": "gzip",
                        "Content-Length": "0",
                        "Host": "149.154.167.51:443",
                        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
                    },
                    "method": "POST",
                    "tor": {
                        "exit_ip": null,
                        "exit_name": null,
                        "is_tor": false
                    },
                    "url": "http://149.154.167.51:443"
                },
                "response": {
                    "body": "<html>\r\n<head><title>501 Not Implemented</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>501 Not Implemented</h1></center>\r\n<hr><center>nginx/0.3.33</center>\r\n</body>\r\n</html>\r\n",
                    "code": 501,
                    "headers": {
                        "Content-Length": "181",
                        "Content-Type": "text/html",
                        "Date": "Mon, 14 Oct 2019 15:06:06 GMT",
                        "Server": "nginx/0.3.33"
                    }
                }
            },
```

I believe this is due to the fact that the URLs you are using to issue requests are missing the trailing `/` at the end of it.

**Comment by hellais** [comment]
Another difference is that the current telegram test doesn't have the port number 80 explicitly listed in the `url` field.

I think that for making it easier to parse measurements we should do the same in `netx`.

See:
`netx`
```
                    "url": "http://149.154.167.91:80"
```

`mk`
```
                    "url": "http://149.154.167.91/"
```

**Comment by bassosimone** [comment]
@hellais thanks for spotting all these differences. I started writing a reply and then I realised it was too long and so I suggest we have a chat over this tomorrow, which seems more efficient to me?

**Comment by bassosimone** [comment]
Discussed all the above with @hellais:

1. is fine but should be documented

2. is easy to fix, probably, but is also to be documented that one should use a URL parser when processing the data, so that URL comparison is better implemented

3. is fine for now, but we need to fix in production

**Comment by bassosimone** [comment]
Superseded by #103
